balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -314,6 +324,71 @@ func deDupAddresses(addrs []resolver.Address) []resolver.Address {
 	return retAddrs
 }
 
+// interleaveAddresses interleaves addresses of both families (IPv4 and IPv6)
+// as per RFC-8305 section 4.
+// Whichever address family is first in the list is followed by an address of
+// the other address family; that is, if the first address in the list is IPv6,
+// then the first IPv4 address should be moved up in the list to be second in
+// the list. It doesn't support configuring "First Address Family Count", i.e.
+// there will always be a single member of the first address family at the
+// beginning of the interleaved list.
+// Addresses that are neither IPv4 nor IPv6 are treated as part of a third
+// "unknown" family for interleaving.
+// See: https://datatracker.ietf.org/doc/html/rfc8305#autoid-6
+func interleaveAddresses(addrs []resolver.Address) []resolver.Address {
+	familyAddrsMap := map[ipAddrFamily][]resolver.Address{}
+	interleavingOrder := []ipAddrFamily{}
+	for _, addr := range addrs {
+		family := addressFamily(addr.Addr)
+		if _, found := familyAddrsMap[family]; !found {
+			interleavingOrder = append(interleavingOrder, family)
+		}
+		familyAddrsMap[family] = append(familyAddrsMap[family], addr)
+	}
+
+	interleavedAddrs := make([]resolver.Address, 0, len(addrs))
+
+	for curFamilyIdx := 0; len(interleavedAddrs) < len(addrs); curFamilyIdx = (curFamilyIdx + 1) % len(interleavingOrder) {
+		// Some IP types may have fewer addresses than others, so we look for
+		// the next type that has a remaining member to add to the interleaved
+		// list.
+		family := interleavingOrder[curFamilyIdx]
+		remainingMembers := familyAddrsMap[family]
+		if len(remainingMembers) > 0 {
+			interleavedAddrs = append(interleavedAddrs, remainingMembers[0])
+			familyAddrsMap[family] = remainingMembers[1:]
+		}
+	}
+
+	return interleavedAddrs
+}
+
+func addressFamily(address string) ipAddrFamily {
+	// Try parsing addresses without a port specified.
+	ip := net.ParseIP(address)
+	if ip == nil {
+		// Try to parse the IP after removing the port.
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return ipAddrFamilyUnknown
+		}
+		ip = net.ParseIP(host)
+	}
+
+	// If using a resolver like passthrough, the hostnames may not be IP
+	// addresses but in some format that the dialer can parse.
+	switch {
+	case ip == nil:
```

Comment:
This could be simplified to:

```go
switch len(ip) {
	case IPv4len:
		return ipAddrFamilyV4
	case IPv6len:
		return ipAddrFamilyV16
	default:
		return ipAddrFamilyUnknown
}
```